node/alphanet/src/lib.rs

```diff
@@ -3,41 +3,110 @@ extern crate serde;
 extern crate serde_derive;
 
 use std::net::{IpAddr, Ipv4Addr, SocketAddr};
-use std::sync::Arc;
+use std::sync::{Arc, RwLock};
 
 use futures::sync::mpsc::channel;
+use futures::Future;
+use tokio::runtime::Runtime;
 
 use client::Client;
 use configs::{get_alphanet_configs, ClientConfig, NetworkConfig, RPCConfig};
 use coroutines::client_task::ClientTask;
+use futures::future::IntoFuture;
+use futures::sink::Sink;
+use futures::stream::TakeWhile;
 use network::spawn_network;
 use nightshade::nightshade_task::spawn_nightshade_task;
 use primitives::types::AccountId;
+use std::thread::JoinHandle;
+use std::time::{Duration, Instant};
+use std::{panic, thread};
+use tokio::prelude::future;
+use tokio::prelude::stream::Stream;
+use tokio::sync::oneshot;
+use tokio::timer::Delay;
+use tokio_signal::unix::Signal;
 use network::proxy::ProxyHandler;
 
 pub fn start() {
     let (client_cfg, network_cfg, rpc_cfg) = get_alphanet_configs();
-    start_from_configs(client_cfg, network_cfg, rpc_cfg);
+    let mut handle = start_from_configs(client_cfg, network_cfg, rpc_cfg);
+    tokio::run(Signal::new(tokio_signal::unix::SIGINT).flatten_stream().into_future().map(drop).map_err(drop));
+    handle.shutdown();
 }
 
 pub fn start_from_configs(
     client_cfg: ClientConfig,
     network_cfg: NetworkConfig,
     rpc_cfg: RPCConfig,
-) {
+) -> ShutdownableThread {
     let client = Arc::new(Client::new(&client_cfg));
     // Use empty pipeline to launch nodes on production.
     let proxy_handlers: Vec<Arc<ProxyHandler>> = vec![];
     start_from_client(client, Some(client_cfg.account_id), network_cfg, rpc_cfg, proxy_handlers)
 }
 
+pub struct ShutdownableThread {
+    thread: Option<JoinHandle<()>>,
+    shutdown_tx: Option<oneshot::Sender<bool>>,
+}
+
+impl ShutdownableThread {
+
+    pub fn start<F>(future: F) -> ShutdownableThread
+    where
+        F: Future<Item = (), Error = ()> + Send + 'static,
+    {
+        let (shutdown_tx, shutdown_rx) = oneshot::channel();
+        let mut runtime = Runtime::new().unwrap();
+        let thread = thread::spawn(|| {
+            runtime.spawn(future);
+            runtime.block_on(
+                shutdown_rx.map_err(|x| ()).into_future()
+            );
+            // TODO HACK: after shutdown all tokio::spawn calls will panic.
+            // for now assume all code with tokio::spawn is exception safe
+            panic::set_hook(Box::new(|_info| {}));
+            runtime.shutdown_now().wait();
+        });
+        let thread = Some(thread);
+        let shutdown_tx = Some(shutdown_tx);
+        ShutdownableThread { thread, shutdown_tx }
+    }
+
+    pub fn shutdown(&mut self) {
```

Comment:
Maybe create a separate type:
```
struct InitializedState {
  thread: JoinHandle<()>,
  oneshot::Sender<bool>
}
```
and then use
```
struct ShutdownableThread {
  state: Option<InitializedState>
}
```
This will avoid having two `take`'s and `expect`.